kolibri/core/content/zip_wsgi.py

```diff
@@ -0,0 +1,448 @@
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import codecs
+import json
+import logging
+import mimetypes
+import os
+import re
+import time
+import zipfile
+from collections import OrderedDict
+
+import html5lib
+from cheroot import wsgi
+from django.conf import settings
+from django.contrib.staticfiles.finders import find
+from django.contrib.staticfiles.storage import staticfiles_storage
+from django.core.cache import cache
+from django.core.handlers.wsgi import WSGIRequest
+from django.http import HttpResponse
+from django.http import HttpResponseNotAllowed
+from django.http import HttpResponseNotFound
+from django.http import HttpResponseNotModified
+from django.http import HttpResponsePermanentRedirect
+from django.http.response import FileResponse
+from django.http.response import StreamingHttpResponse
+from django.template import loader
+from django.utils.cache import patch_response_headers
+from django.utils.encoding import force_str
+from django.utils.http import http_date
+from django.utils.safestring import mark_safe
+
+from kolibri.core.content.errors import InvalidStorageFilenameError
+from kolibri.core.content.utils.paths import get_content_storage_file_path
+from kolibri.core.content.utils.paths import get_hashi_base_path
+from kolibri.core.content.utils.paths import get_hashi_filename
+from kolibri.core.content.utils.paths import get_hashi_path
+from kolibri.core.content.utils.paths import get_zip_content_base_path
+
+
+logger = logging.getLogger(__name__)
+
+
+def add_security_headers(request, response):
+    response["Access-Control-Allow-Origin"] = "*"
+    response["Access-Control-Allow-Methods"] = "GET, OPTIONS"
+    requested_headers = request.META.get("HTTP_ACCESS_CONTROL_REQUEST_HEADERS", "")
+    if requested_headers:
+        response["Access-Control-Allow-Headers"] = requested_headers
+    # restrict CSP to only allow resources to be loaded from self, to prevent info leakage
+    # (e.g. via passing user info out as GET parameters to an attacker's server), or inadvertent data usage
+    response[
+        "Content-Security-Policy"
+    ] = "default-src 'self' 'unsafe-inline' 'unsafe-eval' data: blob:"
```

Comment:
@kollivier we are adding the `blob:` protocol here in the CSP, so I suspect that the Firefox error is not the one causing the issue, but rather related to the Firefox debugger being unable to load the script source when it is a blob.

Similar issue filed here: https://bugzilla.mozilla.org/show_bug.cgi?id=1338558

Not sure what is causing the failed rendering though.

Reply:
The Unity app works for me on the latest Firefox on OS X, so could be a platform-specific or resource issue. Definitely not a general issue with this code at least.